File: frontend/swift/frontend-swift-swiftui/Sources/AppState.swift
import Foundation
import SwiftUI

/// The SPA routes, minus the note screen: `/`, `/login`, `/register`.
enum Screen {
    case home
    case login
    case register
}

enum HealthState {
    case checking
    case ok(status: String, service: String)
    case failed(message: String)
}

/// Screen state + auth flow for the whole cell — the SwiftUI twin of the
/// Compose cell's `AppState`.
@MainActor
final class AppState: ObservableObject {
    private let session: Session
    private let repository: AuthRepository

    @Published private(set) var screen: Screen

    /// `theme: { default: 'dark' }` in `lib/headerConfig.ts`.
    @Published private(set) var isLight = false

    /// `lang: { default: 'en' }`.
    @Published private(set) var lang: Lang = .en

    @Published private(set) var menuOpen = false

    @Published var search = ""
    @Published var menuSearch = ""

    @Published var loginUsername = ""
    @Published var loginPassword = ""
    @Published private(set) var loginError = ""
    @Published private(set) var loginSubmitting = false

    @Published var registerUsername = ""
    @Published var registerPassword = ""
    @Published var registerConfirmPassword = ""
    @Published private(set) var registerError = ""
    @Published private(set) var registerSubmitting = false

    @Published private(set) var welcomeName: String?
    @Published private(set) var health: HealthState = .checking
    @Published var confirmingDelete = false

    init(session: Session = Session()) {
        self.session = session
        repository = AuthRepository(session: session)
        screen = session.token != nil ? .home : .login
    }

    var copy: CopyBook { CopyBook.of(lang) }

    // MARK: - header

    func toggleMenu() { menuOpen.toggle() }

    func closeMenu() { menuOpen = false }

    func toggleLang() { lang = lang.other }

    func toggleTheme() { isLight.toggle() }

    // MARK: - navigation

    func navigate(to target: Screen) {
        closeMenu()
        // `LoginPage`/`RegisterPage` bounce to `/` while a token is stored.
        screen = (target != .home && session.token != nil) ? .home : target
        if screen == .home {
            loadHome()
        }
    }

    /// Escape key / swipe-back equivalent. True when consumed.
    @discardableResult
    func back() -> Bool {
        if menuOpen {
            closeMenu()
            return true
        }
        if confirmingDelete {
            confirmingDelete = false
            return true
        }
        if screen == .register {
            navigate(to: .login)
            return true
        }
        return false
    }

    // MARK: - auth

    func submitLogin() {
        let messages = copy.auth
        let username = loginUsername.trimmingCharacters(in: .whitespacesAndNewlines)
        let password = loginPassword.trimmingCharacters(in: .whitespacesAndNewlines)
        if let validation = validateCredentials(login: username, password: password, copy: messages) {
            loginError = validation
            return
        }
        loginError = ""
        loginSubmitting = true
        Task {
            do {
                let response = try await repository.login(username: username, password: password)
                session.save(response.token)
                loginPassword = ""
                navigate(to: .home)
            } catch {
                loginError = resolveAuthErrorMessage(
                    error,
                    copy: messages,
                    fallback: copy.login.errorWrongCredentials
                )
            }
            loginSubmitting = false
        }
    }

    func submitRegister() {
        let book = copy
        let username = registerUsername.trimmingCharacters(in: .whitespacesAndNewlines)
        let password = registerPassword.trimmingCharacters(in: .whitespacesAndNewlines)
        let confirm = registerConfirmPassword.trimmingCharacters(in: .whitespacesAndNewlines)
        if let validation = validateCredentials(login: username, password: password, copy: book.auth) {
            registerError = validation
            return
        }
        if password != confirm {
            registerError = book.register.errorPasswordMismatch
            return
        }
        registerError = ""
        registerSubmitting = true
        Task {
            do {
                let response = try await repository.register(username: username, password: password)
                session.save(response.token)
                registerPassword = ""
                registerConfirmPassword = ""
                navigate(to: .home)
            } catch {
                registerError = resolveAuthErrorMessage(
                    error,
                    copy: book.auth,
                    fallback: book.register.errorRegistrationFailed
                )
            }
            registerSubmitting = false
        }
    }

    func logout() {
        Task {
            await repository.logout()
            welcomeName = nil
            navigate(to: .login)
        }
    }

    func requestDeleteAccount() { confirmingDelete = true }

    func cancelDelete() { confirmingDelete = false }

    func confirmDeleteAccount() {
        confirmingDelete = false
        Task {
            await repository.deleteAccount()
            welcomeName = nil
            navigate(to: .login)
        }
    }

    /// `HomePage` mount effect: health probe + profile, session dropped on 401.
    func loadHome() {
        health = .checking
        Task {
            do {
                let payload = try await repository.health()
                health = .ok(status: payload.status, service: payload.service)
            } catch {
                let message = (error as? ApiError)?.message ?? ""
                health = .failed(message: message)
            }
        }
        guard session.token != nil else {
            welcomeName = nil
            return
        }
        Task {
            do {
                welcomeName = try await repository.profile().username
            } catch {
                session.clear()
                welcomeName = nil
            }
        }
    }

    var healthText: String {
        switch health {
        case .checking:
            return copy.home.healthChecking
        case let .ok(status, service):
            return formatMessage(
                copy.home.healthOk,
                ["status": status, "service": service, "frontend": uiMount]
            )
        case let .failed(message):
            return formatMessage(copy.home.healthError, ["message": message])
        }
    }

    var healthFailed: Bool {
        if case .failed = health { return true }
        return false
    }
}

File: frontend/kotlin/frontend-kotlin-compose/app/src/main/kotlin/dev/multistack/compose/AppState.kt
package dev.multistack.compose

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import dev.multistack.compose.data.AuthRepository
import dev.multistack.compose.data.Session
import dev.multistack.compose.data.resolveAuthErrorMessage
import dev.multistack.compose.data.validateCredentials
import dev.multistack.compose.i18n.Lang
import dev.multistack.compose.i18n.dictionary
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

/** Product mount id, reported by the Health panel (SPA `UI_MOUNT`). */
const val UI_MOUNT = "frontend-kotlin-compose"

/** The SPA routes, minus the note screen: `/`, `/login`, `/register`. */
enum class Screen { HOME, LOGIN, REGISTER }

sealed interface HealthState {
    data object Checking : HealthState
    data class Ok(val status: String, val service: String) : HealthState
    data class Failed(val message: String) : HealthState
}

/**
 * Screen state + auth flow for the whole cell. Kept outside composition so
 * [MainActivity.dispatchKeyEvent] can close the burger menu on Escape, the way
 * `js/header.js` listens for `keydown`.
 */
class AppState(
    private val session: Session,
    private val repository: AuthRepository,
    private val scope: CoroutineScope,
    private val openUrl: (String) -> Unit,
) {
    var screen by mutableStateOf(if (session.token() != null) Screen.HOME else Screen.LOGIN)
        private set

    /** `theme: { default: 'dark' }` in `lib/headerConfig.ts`. */
    var isLight by mutableStateOf(false)
        private set

    /** `lang: { default: 'en' }`. */
    var lang by mutableStateOf(Lang.EN)
        private set

    var menuOpen by mutableStateOf(false)
        private set

    var search by mutableStateOf("")
    var menuSearch by mutableStateOf("")

    var loginUsername by mutableStateOf("")
    var loginPassword by mutableStateOf("")
    var loginError by mutableStateOf("")
        private set
    var loginSubmitting by mutableStateOf(false)
        private set

    var registerUsername by mutableStateOf("")
    var registerPassword by mutableStateOf("")
    var registerConfirmPassword by mutableStateOf("")
    var registerError by mutableStateOf("")
        private set
    var registerSubmitting by mutableStateOf(false)
        private set

    var welcomeName by mutableStateOf<String?>(null)
        private set
    var health by mutableStateOf<HealthState>(HealthState.Checking)
        private set
    var confirmingDelete by mutableStateOf(false)
        private set

    // ---- header ----------------------------------------------------------

    fun toggleMenu() {
        menuOpen = !menuOpen
    }

    fun closeMenu() {
        menuOpen = false
    }

    fun toggleLang() {
        lang = lang.other()
    }

    fun toggleTheme() {
        isLight = !isLight
    }

    /** `header-nav-stack` → `/stack/` board. No WebView: hand off to the browser. */
    fun openStackIndex() {
        openUrl(BuildConfig.STACK_INDEX_URL)
    }

    // ---- navigation ------------------------------------------------------

    fun navigate(target: Screen) {
        closeMenu()
        // `LoginPage`/`RegisterPage` bounce to `/` while a token is stored.
        screen = if (target != Screen.HOME && session.token() != null) Screen.HOME else target
        if (screen == Screen.HOME) {
            loadHome()
        }
    }

    /** Android back / Appium `pressKeyCode(ESCAPE)`. True when consumed. */
    fun back(): Boolean {
        if (menuOpen) {
            closeMenu()
            return true
        }
        if (confirmingDelete) {
            cancelDelete()
            return true
        }
        if (screen == Screen.REGISTER) {
            navigate(Screen.LOGIN)
            return true
        }
        return false
    }

    // ---- auth ------------------------------------------------------------

    fun submitLogin() {
        val messages = dictionary(lang).auth
        val username = loginUsername.trim()
        val password = loginPassword.trim()
        validateCredentials(username, password, messages)?.let {
            loginError = it
            return
        }
        loginError = ""
        loginSubmitting = true
        scope.launch {
            try {
                val response = repository.login(username, password)
                session.save(response.token)
                loginPassword = ""
                navigate(Screen.HOME)
            } catch (error: Exception) {
                loginError = resolveAuthErrorMessage(
                    error,
                    messages,
                    dictionary(lang).login.errorWrongCredentials,
                )
            } finally {
                loginSubmitting = false
            }
        }
    }

    fun submitRegister() {
        val copy = dictionary(lang)
        val username = registerUsername.trim()
        val password = registerPassword.trim()
        val confirm = registerConfirmPassword.trim()
        validateCredentials(username, password, copy.auth)?.let {
            registerError = it
            return
        }
        if (password != confirm) {
            registerError = copy.register.errorPasswordMismatch
            return
        }
        registerError = ""
        registerSubmitting = true
        scope.launch {
            try {
                val response = repository.register(username, password)
                session.save(response.token)
                registerPassword = ""
                registerConfirmPassword = ""
                navigate(Screen.HOME)
            } catch (error: Exception) {
                registerError = resolveAuthErrorMessage(
                    error,
                    copy.auth,
                    copy.register.errorRegistrationFailed,
                )
            } finally {
                registerSubmitting = false
            }
        }
    }

    fun logout() {
        scope.launch {
            repository.logout()
            welcomeName = null
            navigate(Screen.LOGIN)
        }
    }

    fun requestDeleteAccount() {
        confirmingDelete = true
    }

    fun cancelDelete() {
        confirmingDelete = false
    }

    fun confirmDeleteAccount() {
        confirmingDelete = false
        scope.launch {
            repository.deleteAccount()
            welcomeName = null
            navigate(Screen.LOGIN)
        }
    }

    /** `HomePage` mount effect: health probe + profile, session dropped on 401. */
    fun loadHome() {
        health = HealthState.Checking
        scope.launch {
            health = try {
                val payload = repository.health()
                HealthState.Ok(status = payload.status, service = payload.service)
            } catch (error: Exception) {
                HealthState.Failed(error.message.orEmpty())
            }
        }
        if (session.token() == null) {
            welcomeName = null
            return
        }
        scope.launch {
            welcomeName = try {
                repository.profile().username
            } catch (_: Exception) {
                session.clear()
                null
            }
        }
    }
}
